Sources/SwiftSyntax/SourceEdit.swift

```diff
@@ -13,24 +13,68 @@
 /// A textual edit to the original source represented by a range and a
 /// replacement.
 public struct SourceEdit: Equatable, Sendable {
-  /// The half-open range that this edit applies to.
-  public let range: Range<AbsolutePosition>
-  /// The text to replace the original range with. Empty for a deletion.
-  public let replacement: String
-
-  /// Length of the original source range that this edit applies to. Zero if
-  /// this is an addition.
-  public var length: SourceLength {
-    return SourceLength(utf8Length: range.upperBound.utf8Offset - range.lowerBound.utf8Offset)
+  /// The byte range of the original source buffer that the edit applies to.
+  public let range: ByteSourceRange
+
+  /// The UTF-8 bytes that should be inserted as part of the edit
+  public let replacement: [UInt8]
+
+  /// A string representation of the replacement
+  public var replacementText: String { return String(decoding: replacement, as: UTF8.self) }
+
+  /// The length of the edit replacement in UTF8 bytes.
+  public var replacementLength: Int { replacement.count }
+
+  public var offset: Int { return range.offset }
+
+  public var length: Int { return range.length }
+
+  public var endOffset: Int { return range.endOffset }
+
+  /// After the edit has been applied the range of the replacement text.
+  public var replacementRange: ByteSourceRange {
+    return ByteSourceRange(offset: offset, length: replacementLength)
+  }
+
+  @available(*, deprecated, message: "Use IncrementalEdit(range:replacement:) instead")
+  public init(range: ByteSourceRange, replacementLength: Int) {
+    self.range = range
+    self.replacement = Array(repeating: UInt8(ascii: " "), count: replacementLength)
+  }
+
+  @available(*, deprecated, message: "Use IncrementalEdit(offset:length:replacement:) instead")
+  public init(offset: Int, length: Int, replacementLength: Int) {
+    self.range = ByteSourceRange(offset: offset, length: length)
+    self.replacement = Array(repeating: UInt8(ascii: " "), count: replacementLength)
   }
 
   /// Create an edit to replace `range` in the original source with
   /// `replacement`.
   public init(range: Range<AbsolutePosition>, replacement: String) {
-    self.range = range
+    self.range = ByteSourceRange(
+      offset: range.lowerBound.utf8Offset,
+      length: range.upperBound.utf8Offset - range.lowerBound.utf8Offset
+    )
+    self.replacement = Array(replacement.utf8)
+  }
+
+  public init(offset: Int, length: Int, replacement: [UInt8]) {
+    self.range = ByteSourceRange(offset: offset, length: length)
     self.replacement = replacement
   }
 
+  public init(offset: Int, length: Int, replacement: String) {
+    self.init(offset: offset, length: length, replacement: Array(replacement.utf8))
+  }
```

Comment:
I would like to deprecate the initializers that take an `Int` because it’s very easy to mix up UTF-8, UTF-16 and character offsets into a source file. That’s why I prefer to use `AbsolutePosition` and `SourceLength`.